src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,405 @@
+            if (length > (hmacLen * 255)) {
+                throw new InvalidAlgorithmParameterException(
+                    "Requested length exceeds maximum allowed length");
+            }
+            // perform extract and then expand
+            try {
+                pseudoRandomKey = hkdfExtract(inputKeyMaterial, salt);
+                return hkdfExpand(pseudoRandomKey, info, length);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidAlgorithmParameterException(
+                    "an HKDF ExtractThenExpand could not be initialized with "
+                    + "the given key or salt material", ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        }
+        throw new InvalidAlgorithmParameterException(
+            "an HKDF derivation requires a valid HKDFParameterSpec");
+    }
+
+    private byte[] consolidateKeyMaterial(List<SecretKey> keys)
+        throws InvalidKeyException {
+        if (keys != null && !keys.isEmpty()) {
+            ArrayList<SecretKey> localKeys = new ArrayList<>(keys);
+            if (localKeys.size() == 1) {
+                // return this element
+                SecretKey checkIt = localKeys.get(0);
+                return CipherCore.getKeyBytes(checkIt);
+            } else {
+                ByteArrayOutputStream os = new ByteArrayOutputStream();
+                for (SecretKey workItem : localKeys) {
+                    os.writeBytes(CipherCore.getKeyBytes(workItem));
+                }
+                // deliberately omitting os.flush(), since we are writing to
+                // memory, and toByteArray() reads like there isn't an explicit
+                // need for this call
+                return os.toByteArray();
+            }
+        } else if(keys != null) {
+                return null;
+        } else {
+            throw new InvalidKeyException(
+                "List of key segments could not be consolidated");
+        }
+    }
+
+    /**
+     * Perform the HKDF-Extract operation.
+     *
+     * @param inputKeyMaterial
+     *     the input keying material used for the HKDF-Extract operation.
+     * @param salt
+     *     the salt value used for HKDF-Extract; {@code null} if no salt value
+     *     is provided.
+     *
+     * @return a byte array containing the pseudorandom key (PRK)
+     *
+     * @throws InvalidKeyException
+     *     if an invalid salt was provided through the
+     *     {@code HkdfParameterSpec}
+     */
+    private byte[] hkdfExtract(byte[] inputKeyMaterial, byte[] salt)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+
+        if (salt == null) {
```

Comment:
I understand `addSalt(new byte[0])` is not a problem, but what about
```
addSalt(new SecretKey() {
    public byte[] getEncoded() { return new byte[0]; }
    ....
};
```
Anyway, with the new length check, this will be detected.